File: 실버/11724(연결 요소의 개수)/Main.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Queue;

class Main {
	static boolean[] visited;
	static int N, M;
	static ArrayList<ArrayList<Integer>> graph;
	public static void main(String[] args) throws IOException{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		String[] temp = br.readLine().split(" ");
		N = Integer.parseInt(temp[0]);
		M = Integer.parseInt(temp[1]);

		// 2차원 arrayList 초기화
		graph = new ArrayList<>();
		for(int i=0; i<N+1; i++)
			graph.add(new ArrayList<Integer>());

		// 값 입력받기
		for(int i=0; i<M; i++){
			temp = br.readLine().split(" ");
			int a = Integer.parseInt(temp[0]);
			int b = Integer.parseInt(temp[1]);

			graph.get(a).add(b);
			graph.get(b).add(a);
		}

		// 방문
		int cnt = 0;
		visited = new boolean[N+1];
		for(int i=1; i<=N; i++){
			if(!visited[i]){
				bfs(i);
				cnt++;
			}
		}
		System.out.println(cnt);
		br.close();
	}

	public static void bfs(int start){
		Queue<Integer> q = new LinkedList<>();
		visited[start] = true;

		q.offer(start);
		while(!q.isEmpty()){
			int p = q.poll();
			
			// 연결된 점들 모두 탐색
			// 방문하지 않았다면 방문하고 다음 번 탐색을 위해 queue에 넣기
			for(int b : graph.get(p)){
				if(!visited[b]){
					visited[b] = true;
					q.offer(b);
				}
			}
		}
	}
}

File: 실버/11724(연결 요소의 개수)/Main.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include <queue>
using namespace std;

int N, M;
bool visited[1001];
vector<vector<int>> graph;

void bfs(int start) {
	queue<int> q;

	q.push(start);
	visited[start] = true;

	while (!q.empty()) {
		int a = q.front();	// q의 맨 앞에서 꺼낼 값
		q.pop();	// 실제 꺼내기

		vector<int>::iterator it;
		for (it = graph[a].begin(); it != graph[a].end(); it++) {
			if (!visited[*it]) {
				visited[*it] = true;
				q.push(*it);
			}
		}
	}
}

int main() {
	// IO 속도 향상
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);
	
	cin >> N >> M;
	// 2차원 벡터 초기화
	for (int i = 0; i <= N; i++) {
		vector<int> temp;
		graph.push_back(temp);
	}

	// 값 입력 받기
	for (int i = 0; i < M; i++) {
		int a, b;
		cin >> a >> b;
		// 그래프 저장
		graph.at(a).push_back(b);
		graph.at(b).push_back(a);
	}
	
	// 연결 요소 탐색
	int cnt = 0;
	for (int i = 1; i <= N; i++) {
		if (!visited[i]) {
			bfs(i);
			cnt++;
		}
	}
	
	cout << cnt << endl;
	return 0;
}